### PR TITLE
Docs: how to enable custom tools guide

### PR DESCRIPTION
**Pull request by itisallgood**
https://docs.robusta.dev/holmes-docs-enable-custom-tools/configuration/ai-analysis.html#toolsets-examples

### CODE REVIEW

File: docs/configuration/ai-analysis.rst [hunk 1]
```diff
@@ -299,36 +299,318 @@ Advanced - Customizing HolmesGPT
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 
-Adding Custom Tools to Holmes
+Holmes Toolsets
 -------------------------------------
 
-Holmes allows you to define custom toolsets that enhance its functionality by enabling additional tools to run Kubernetes commands or other tasks.
+Holmes allows you to define tools that enhance its functionality by enabling additional tools to run Kubernetes commands or other tasks.
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:305]: This is a bit repetitive and doesn't say much. I suggest

> Holmes allows you to define integrations (toolsets) that fetch data from external sources. This data will be automatically used in investigations when relevant.

File: docs/configuration/ai-analysis.rst [hunk 2]
```diff
@@ -299,36 +299,318 @@ Advanced - Customizing HolmesGPT
...
+
+Default Toolsets in Holmes
+--------------------------
+Holmes comes with a set of default toolsets that provide tools out-of-the-box. Some of these toolsets are enabled by default such as interacting with Kubernetes resources, or fetching logs and metrics.
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:309]: Holmes comes with a set of builtin toolsets. Most of these toolsets are enabled by default, such as toolsets to read Kubernetes resources and fetch logs. Some builtin toolsets are disabled by default and can be enabled by the user by providing credentials or API keys to external systems.

File: docs/configuration/ai-analysis.rst [hunk 3]
```diff
@@ -299,36 +299,318 @@ Advanced - Customizing HolmesGPT
...
+Full list can be found `here <https://github.com/robusta-dev/holmesgpt/tree/master/holmes/plugins/toolsets>`_
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:310]: The full list...

File: docs/configuration/ai-analysis.rst [hunk 4]
```diff
@@ -299,36 +299,318 @@ Advanced - Customizing HolmesGPT
...
 
-In this guide, we will show how to add a custom toolset to Holmes in your ``generated_values.yaml`` file.
+By default, any of the toolsets may be disabled. You can enable or disable them by modifying the ``generated_values.yaml`` file.
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:312]: I suggest replacing with:

> You can enable or disable toolsets with the `holmes.toolsets` Helm value:

File: docs/configuration/ai-analysis.rst [hunk 5]
```diff
@@ -299,36 +299,318 @@ Advanced - Customizing HolmesGPT
...
 
 .. code-block:: yaml
+    enableHolmesGPT: true
+    holmes:
+      additionalEnvVars:
+        - name: ROBUSTA_AI
+          value: "true"
+      toolsets:
+        kubernetes/logs:
+          enabled: false
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:322]: add a comment here

```
# disable the builtin kubernetes/logs toolset - e.g. if you want Holmes to only read logs from Loki instead (requires enabling a loki toolset)
```

File: docs/configuration/ai-analysis.rst [hunk 6]
```diff
@@ -299,36 +299,318 @@ Advanced - Customizing HolmesGPT
...
+
+After making changes, apply them using Helm:
+
+.. code-block:: bash
 
+    helm upgrade robusta robusta/robusta --values=generated_values.yaml --set clusterName=<YOUR_CLUSTER_NAME>
+
+
+Additionally, you can overwrite any of the fields in the default toolsets, such as updating the tools, prerequisites, or descriptions, to customize them for your specific needs.
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:331]: You can override fields from the default toolsets to customize them for your needs.

File: docs/configuration/ai-analysis.rst [hunk 7]
```diff
@@ -299,36 +299,318 @@ Advanced - Customizing HolmesGPT
...
-          # List of tools the LLM can use - this is the important part
-          tools:
-          # Name is a unique identifier for the tool
-            - name: "explain_resource"
-              # The LLM looks at this description when deciding what tools are relevant for each task
-              description: "Provides detailed explanation of Kubernetes resources using kubectl explain"
-              # A templated bash command using Jinja2 templates
-              # The LLM can only control parameters that you expose as template variables like {{ resource_name }}
-              command: "kubectl explain {{ resource_name }}"
+      kubernetes/logs:
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:341]: Lets show a real example here. @arikalon1 any ideas on something useful to show here?

File: docs/configuration/ai-analysis.rst [hunk 8]
```diff
@@ -299,36 +299,318 @@ Advanced - Customizing HolmesGPT
...
+        description: "My custom description for default toolset."
+
+
+How to define a toolset?
+-------------------------------------
+A toolset is defined in ``generated_values.yaml``. Each toolset has a unique name and has to contain tools.
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:347]: > A toolset is defined in your Helm values (``generated_values.yaml``). Each toolset has a unique name and has to contain tools.

File: docs/configuration/ai-analysis.rst [hunk 9]
```diff
@@ -299,36 +299,318 @@ Advanced - Customizing HolmesGPT
...
-              # The LLM looks at this description when deciding what tools are relevant for each task
-              description: "Provides detailed explanation of Kubernetes resources using kubectl explain"
-              # A templated bash command using Jinja2 templates
-              # The LLM can only control parameters that you expose as template variables like {{ resource_name }}
-              command: "kubectl explain {{ resource_name }}"
+      kubernetes/logs:
+        description: "My custom description for default toolset."
+
+
+How to define a toolset?
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:345]: How to define a new toolset

File: docs/configuration/ai-analysis.rst [hunk 10]
```diff
@@ -299,36 +299,318 @@ Advanced - Customizing HolmesGPT
...
+Full list can be found `here <https://github.com/robusta-dev/holmesgpt/tree/master/holmes/plugins/toolsets>`_
 
-In this guide, we will show how to add a custom toolset to Holmes in your ``generated_values.yaml`` file.
+By default, any of the toolsets may be disabled. You can enable or disable them by modifying the ``generated_values.yaml`` file.
 
 .. code-block:: yaml
+    enableHolmesGPT: true
+    holmes:
+      additionalEnvVars:
+        - name: ROBUSTA_AI
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:318]: remove the additionalEnvVars section - it isn't relevant to the example

File: docs/configuration/ai-analysis.rst [hunk 11]
```diff
@@ -299,36 +299,318 @@ Advanced - Customizing HolmesGPT
...
+     - boolean
+     - Indicates whether the toolset is enabled. Defaults to ``true``. If set to ``false``, the toolset will be disabled.
+     - No
+   * - ``name``
+     - string
+     - A unique identifier for the toolset. Used for informational purposes and logging.
+     - **Yes**
+   * - ``description``
+     - string
+     - A brief description of the toolset's purpose. Helps users understand what the toolset does.
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:401]: Is this seen by the LLM itself? If not, be explicit about that.

File: docs/configuration/ai-analysis.rst [hunk 12]
```diff
@@ -299,36 +299,318 @@ Advanced - Customizing HolmesGPT
...
+     - list
+     - Tags for categorizing toolsets, ``core`` will be used for all Holmes features (both cli's commands and chats in UI). The ``cluster`` tag is used for UI functionality, while ``cli`` is for for command-line specific tools. Default to ``[core,]``.
+     - No
+   * - ``installation_instructions``
+     - string
+     - Instructions on how to install prerequisites required by the toolset.
+     - No
+   * - ``variables``
+     - dictionary
+     - A set of key-value pairs defining variables that can be used within tools and commands. Values can reference environment variables using the ``$VARIABLE_NAME`` syntax.
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:421]: I would be explicit here that these variables are set by the user, not the LLM, and are typical used for configuring API Keys and credentials that the LLM does not see.

File: docs/configuration/ai-analysis.rst [hunk 13]
```diff
@@ -299,36 +299,318 @@ Advanced - Customizing HolmesGPT
...
+     - **Yes**
+   * - ``command``
+     - string
+     - A shell command template that the tool will execute. Can include variables and parameters using Jinja2 syntax (``{{ variable_name }}``).
+     - At least one of ``command`` or ``script`` is required
+   * - ``script``
+     - string
+     - The content of a script that the tool will execute. Use this if your tool requires a multi-line script.
+     - At least one of ``command`` or ``script`` is required
+   * - ``parameters``
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:462]: I would be explicit that these parameters are filled in by the LLM

File: docs/configuration/ai-analysis.rst [hunk 14]
```diff
@@ -299,36 +299,318 @@ Advanced - Customizing HolmesGPT
...
+     - dictionary
+     - Defines the inputs required for the tool. Each parameter has its own fields.
+     - No
+   * - ``additional_instructions``
+     - string
+     - Additional shell commands or processing instructions applied to the output of this tool.
+     - No
+
+**Parameter Fields (Within `parameters`, not required)**
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:471]: Explain the not required (if missing we infer it)

File: docs/configuration/ai-analysis.rst [hunk 15]
```diff
@@ -299,36 +299,318 @@ Advanced - Customizing HolmesGPT
...
+   * - ``required``
+     - boolean
+     - Indicates whether the parameter is required. Defaults to ``true``.
+     - No
+
+
+Toolsets Examples
+-----------------
+
+**Example 1: AWS S3 Management Toolset**
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:498]: Title doesn't match the toolset itself (title is about AWS S3)

File: docs/configuration/ai-analysis.rst [hunk 16]
```diff
@@ -299,36 +299,318 @@ Advanced - Customizing HolmesGPT
...
+            - cluster
+          prerequisites:
+            - command: "curl -o /dev/null -s -w '%{http_code}' https://example.com "
+              expected_output: "200"
+          tools:
+            - name: "curl_example"
+              description: "Perform a GET request to example.com"
+              command: "curl -X GET https://example.com"
+            - name: "curl_with_params"
+              description: "Perform a GET request to example.com with query parameters"
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:606]: I would give a description that helps the LLM understand when to use this (use this to xyz...)

File: docs/configuration/ai-analysis.rst [hunk 17]
```diff
@@ -299,36 +299,318 @@ Advanced - Customizing HolmesGPT
...
+
+Adding Custom Tools to Holmes
+-----------------------------
+As an example, let's add custom toolset named ``http_tools`` that  makes requests to ``example.com``
+
+.. code-block:: yaml
+
+    enableHolmesGPT: true
+    holmes:
+      additionalEnvVars:
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:587]: remove the additionalEnvVars section - it isn't relevant to the example

File: docs/configuration/ai-analysis.rst [hunk 18]
```diff
@@ -299,36 +299,318 @@ Advanced - Customizing HolmesGPT
...
+              kubectl exec -n {{ namespace }} {{ pod_name }} -- du -sh /
+          - name: "kubectl_find_evicted_pods"
+            description: "List all evicted pods in a specific namespace."
+            command: "kubectl get pods -n {{ namespace }} --field-selector=status.phase=Failed | grep Evicted"
+          - name: "kubectl_drain_node"
+            description: "Drain a node safely by evicting all pods."
+            command: "kubectl drain {{ node_name }} --ignore-daemonsets --force --delete-emptydir-data"
+
+
+Adding Custom Tools to Holmes
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:579]: Is this the same as the two examples above?

**Review by aantn — CHANGES_REQUESTED** [review]
Left some comments on docs

File: docs/configuration/ai-analysis.rst (hunk 11 above)
Comment (itisallgood) [reply docs/configuration/ai-analysis.rst:401]: Yes. Updated the field's description

File: docs/configuration/ai-analysis.rst (hunk 7 above)
Comment (itisallgood) [reply docs/configuration/ai-analysis.rst:341]: I added more realistic example with enhanced functionality of confluence

File: docs/configuration/ai-analysis.rst (hunk 18 above)
Comment (itisallgood) [reply docs/configuration/ai-analysis.rst:579]: Basically yes, but previous toolsets were just examples, but here we give a short guide on how to add a toolset

File: docs/configuration/ai-analysis.rst (hunk 13 above)
Comment (itisallgood) [reply docs/configuration/ai-analysis.rst:462]: Updated description

File: docs/configuration/ai-analysis.rst (hunk 16 above)
Comment (itisallgood) [reply docs/configuration/ai-analysis.rst:606]: Updated descriptions

File: docs/configuration/ai-analysis.rst (hunk 18 above)
Comment (aantn) [reply docs/configuration/ai-analysis.rst:589]: Can you give full insttructions on the previous 2 examples as well? (And the exact yaml structure necessary - e.g. `holmes.toolsets` and not just `toolsets`)

So we will have

Example 1 ...
Example 2 ...
Example 3 ...

And each will show the full instructions

File: docs/configuration/ai-analysis.rst [hunk 19]
```diff
@@ -299,36 +299,320 @@ Advanced - Customizing HolmesGPT
...
+     - A unique identifier for the tool within the toolset.
+     - **Yes**
+   * - ``description``
+     - string
+     - A brief description of the tool's purpose. Helps Holmes decide when to use this tool.
+     - **Yes**
+   * - ``command``
+     - string
+     - A shell command template that the tool will execute. Can include variables and parameters using Jinja2 syntax (``{{ variable_name }}``).
+     - At least one of ``command`` or ``script`` is required
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:467]: This isn't accurate. It's not *at least one of*. It's *either*
Please update below as well

File: docs/configuration/ai-analysis.rst [hunk 20]
```diff
@@ -299,36 +299,320 @@ Advanced - Customizing HolmesGPT
...
+   * - ``script``
+     - string
+     - The content of a script that the tool will execute. Use this if your tool requires a multi-line script.
+     - At least one of ``command`` or ``script`` is required
+   * - ``parameters``
+     - dictionary
+     - Defines the inputs required for the tool. These parameters are filled in by the LLM based on the context of the task, allowing dynamic customization of the tool's execution. Each parameter has its own fields, such as type, description, and whether it is required.
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:474]: Mention here that this is optional and can be infered

File: docs/configuration/ai-analysis.rst [hunk 21]
```diff
@@ -299,36 +299,320 @@ Advanced - Customizing HolmesGPT
...
+     - No
+   * - ``additional_instructions``
+     - string
+     - Additional shell commands or processing instructions applied to the output of this tool.
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:478]: Can you give an example of how this would be used?

File: docs/configuration/ai-analysis.rst [hunk 22]
```diff
@@ -299,36 +299,320 @@ Advanced - Customizing HolmesGPT
...
+            parameters:
+              <parameter_name>:
+                type: "<string>"
+                description: "<string>"
+                required: <true|false>
+
+Toolset Fields
+--------------
+
+.. list-table::
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:393]: this doesn't render right in the preview url - see https://docs.robusta.dev/holmes-docs-enable-custom-tools/configuration/ai-analysis.html#how-to-define-a-new-toolset

File: docs/configuration/ai-analysis.rst [hunk 23]
```diff
@@ -299,36 +299,320 @@ Advanced - Customizing HolmesGPT
...
+        tags:
+          - cli
+        variables:
+          github_token: "$GITHUB_TOKEN"
+        prerequisites:
+          - env:
+            - "GITHUB_TOKEN"
+          - command: "curl --version"
+        tools:
+          - name: "list_user_repos"
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:527]: explain in a comment that parameters here are inferred

File: docs/configuration/ai-analysis.rst [hunk 24]
```diff
@@ -299,36 +299,320 @@ Advanced - Customizing HolmesGPT
...
+            description: "Lists all repositories for a GitHub user"
+            command: "curl -H 'Authorization: token {{ github_token }}' https://api.github.com/users/{{ username }}/repos"
+
+          - name: "show_recent_commits"
+            description: "Shows the most recent commits for a repository"
+            command: "cd {{ repo_dir }} && git log -{{number_of_commits}} --oneline"
+
+          - name: "get_repo_details"
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:535]: explain in a comment that here you chose to be explicit about parameters - maybe show an example where this is actually useful - e.g. if you add a parameter that you want to be optional

File: docs/configuration/ai-analysis.rst [hunk 25]
```diff
@@ -299,36 +299,320 @@ Advanced - Customizing HolmesGPT
...
+        variables:
+          github_token: "$GITHUB_TOKEN"
+        prerequisites:
+          - env:
+            - "GITHUB_TOKEN"
+          - command: "curl --version"
+        tools:
+          - name: "list_user_repos"
+            description: "Lists all repositories for a GitHub user"
+            command: "curl -H 'Authorization: token {{ github_token }}' https://api.github.com/users/{{ username }}/repos"
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:529]: do you need to override the user_description so that the github_token is not shown there?

**Review by aantn — CHANGES_REQUESTED** [review]
Looking good. Left a few more comments. This should be the final review, I hope.

File: docs/configuration/ai-analysis.rst (hunk 21 above)
Comment (itisallgood) [reply docs/configuration/ai-analysis.rst]: Updated the description

File: docs/configuration/ai-analysis.rst (hunk 23 above)
Comment (itisallgood) [reply docs/configuration/ai-analysis.rst]: Added comments

File: docs/configuration/ai-analysis.rst (hunk 25 above)
Comment (itisallgood) [reply docs/configuration/ai-analysis.rst]: This was changed as we removed variables section

File: docs/configuration/ai-analysis.rst [hunk 26]
```diff
@@ -299,44 +299,387 @@ Advanced - Customizing HolmesGPT
...
+     - No
+   * - ``tags``
+     - list
+     - Tags for categorizing toolsets, ``core`` will be used for all Holmes features (both cli's commands and chats in UI). The ``cluster`` tag is used for UI functionality, while ``cli`` is for for command-line specific tools. Default to ``[core,]``.
+     - No
+   * - ``installation_instructions``
+     - string
+     - Instructions on how to install prerequisites required by the toolset.
+     - No
+   * - ``variables``
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:429]: This is no longer relevant, right?

File: docs/configuration/ai-analysis.rst [hunk 27]
```diff
@@ -299,44 +299,387 @@ Advanced - Customizing HolmesGPT
...
+     - string
+     - A unique identifier for the tool within the toolset.
+     - **Yes**
+   * - ``description``
+     - string
+     - A brief description of the tool's purpose. Helps Holmes decide when to use this tool.
+     - **Yes**
+   * - ``command``
+     - string
+     - A shell command template that the tool will execute. Can include variables and parameters using Jinja2 syntax (``{{ variable_name }}``).
```
Comment (aantn) [inline docs/configuration/ai-analysis.rst:467]: Can include environment variables using `${ENV_VAR}` or parameters that the LLM will fill in, using Jinja2 syntax (``{{ param_name }}``)

**Review by aantn — APPROVED** [review]
Left some final comments. Please fix then merge. No need for a further review.
Thanks!

**Review by moshemorad — APPROVED** [review]

**Pull request merged**